presto-hive/src/main/java/com/facebook/presto/hive/rule/HiveParquetDereferencePushDown.java

```diff
@@ -0,0 +1,403 @@
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.rowExpressionService = requireNonNull(rowExpressionService, "rowExpressionService is null");
+    }
+
+    private static Map<RowExpression, NestedColumn> extractDereferences(ConnectorSession session, ExpressionOptimizer expressionOptimizer, Set<RowExpression> expressions)
+    {
+        Set<RowExpression> dereferenceAndVariableExprs = new HashSet<>();
+        expressions.forEach(expr -> expr.accept(new ExtractDereferenceAndVariables(session, expressionOptimizer), dereferenceAndVariableExprs));
+
+        // keep prefix only expressions
+        List<RowExpression> dereferences = dereferenceAndVariableExprs.stream()
+                .filter(expression -> !prefixExists(expression, dereferenceAndVariableExprs))
+                .filter(expression -> expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm() == DEREFERENCE)
+                .collect(Collectors.toList());
+
+        return dereferences.stream().collect(toMap(identity(), dereference -> createNestedColumn(dereference, expressionOptimizer, session)));
+    }
+
+    private static boolean prefixExists(RowExpression expression, Set<RowExpression> allExpressions)
+    {
+        int[] referenceCount = {0};
+        expression.accept(
+                new DefaultRowExpressionTraversalVisitor<int[]>()
+                {
+                    @Override
+                    public Void visitSpecialForm(SpecialFormExpression specialForm, int[] context)
+                    {
+                        if (specialForm.getForm() != DEREFERENCE) {
+                            return super.visitSpecialForm(specialForm, context);
+                        }
+
+                        if (allExpressions.contains(specialForm)) {
+                            referenceCount[0] += 1;
+                        }
+
+                        RowExpression base = specialForm.getArguments().get(0);
+                        base.accept(this, context);
+                        return null;
+                    }
+
+                    @Override
+                    public Void visitVariableReference(VariableReferenceExpression reference, int[] context)
+                    {
+                        if (allExpressions.contains(reference)) {
+                            referenceCount[0] += 1;
+                        }
+                        return null;
+                    }
+                }, referenceCount);
+
+        return referenceCount[0] > 1;
+    }
+
+    private static NestedColumn createNestedColumn(RowExpression rowExpression, ExpressionOptimizer expressionOptimizer, ConnectorSession session)
+    {
+        if (!(rowExpression instanceof SpecialFormExpression) || ((SpecialFormExpression) rowExpression).getForm() != DEREFERENCE) {
+            throw new IllegalArgumentException("invalid expression: " + rowExpression);
```

Comment:
improve the error message with "expecting dereference expression, of type SpecialFormExpression with form DEREFERENCE"